File: multiple-languages/dotnet/AlibabaCloud.SDK.ROS.CDK.Rocketmq5/AlibabaCloud/SDK/ROS/CDK/Rocketmq5/Acl.cs
using Amazon.JSII.Runtime.Deputy;

#pragma warning disable CS0672,CS0809,CS1591

namespace AlibabaCloud.SDK.ROS.CDK.Rocketmq5
{
    /// <summary>This class encapsulates and extends the ROS resource type `ALIYUN::ROCKETMQ5::Acl`.</summary>
    /// <remarks>
    /// <strong>Note</strong>: This class may have some new functions to facilitate development, so it is recommended to use this class instead of `RosAcl`for a more convenient development experience.
    /// See https://www.alibabacloud.com/help/ros/developer-reference/aliyun-rocketmq5-acl
    /// </remarks>
    [JsiiClass(nativeType: typeof(AlibabaCloud.SDK.ROS.CDK.Rocketmq5.Acl), fullyQualifiedName: "@alicloud/ros-cdk-rocketmq5.Acl", parametersJson: "[{\"name\":\"scope\",\"type\":{\"fqn\":\"@alicloud/ros-cdk-core.Construct\"}},{\"name\":\"id\",\"type\":{\"primitive\":\"string\"}},{\"name\":\"props\",\"type\":{\"fqn\":\"@alicloud/ros-cdk-rocketmq5.AclProps\"}},{\"name\":\"enableResourcePropertyConstraint\",\"optional\":true,\"type\":{\"primitive\":\"boolean\"}}]")]
    public class Acl : AlibabaCloud.SDK.ROS.CDK.Core.Resource_, AlibabaCloud.SDK.ROS.CDK.Rocketmq5.IAcl
    {
        /// <summary>Param scope - scope in which this resource is defined Param id    - scoped id of the resource Param props - resource properties.</summary>
        public Acl(AlibabaCloud.SDK.ROS.CDK.Core.Construct scope, string id, AlibabaCloud.SDK.ROS.CDK.Rocketmq5.IAclProps props, bool? enableResourcePropertyConstraint = null): base(_MakeDeputyProps(scope, id, props, enableResourcePropertyConstraint))
        {
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        private static DeputyProps _MakeDeputyProps(AlibabaCloud.SDK.ROS.CDK.Core.Construct scope, string id, AlibabaCloud.SDK.ROS.CDK.Rocketmq5.IAclProps props, bool? enableResourcePropertyConstraint = null)
        {
            return new DeputyProps(new object?[]{scope, id, props, enableResourcePropertyConstraint});
        }

        /// <summary>Used by jsii to construct an instance of this class from a Javascript-owned object reference</summary>
        /// <param name="reference">The Javascript-owned object reference</param>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        protected Acl(ByRefValue reference): base(reference)
        {
        }

        /// <summary>Used by jsii to construct an instance of this class from DeputyProps</summary>
        /// <param name="props">The deputy props</param>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        protected Acl(DeputyProps props): base(props)
        {
        }

        /// <summary>Attribute Actions: The type of operation authorized.</summary>
        [JsiiProperty(name: "attrActions", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public virtual object AttrActions
        {
            get => GetInstanceProperty<object>()!;
        }

        /// <summary>Attribute Decision: The decision result of the authorization.</summary>
        [JsiiProperty(name: "attrDecision", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public virtual object AttrDecision
        {
            get => GetInstanceProperty<object>()!;
        }

        /// <summary>Attribute InstanceId: The ID of the RocketMQ instance.</summary>
        [JsiiProperty(name: "attrInstanceId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public virtual object AttrInstanceId
        {
            get => GetInstanceProperty<object>()!;
        }

        /// <summary>Attribute IpWhitelists: The IP addresses in the whitelist.</summary>
        [JsiiProperty(name: "attrIpWhitelists", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public virtual object AttrIpWhitelists
        {
            get => GetInstanceProperty<object>()!;
        }

        /// <summary>Attribute ResourceName: The name of the resource on which you want to grant permissions.</summary>
        [JsiiProperty(name: "attrResourceName", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public virtual object AttrResourceName
        {
            get => GetInstanceProperty<object>()!;
        }

        /// <summary>Attribute Username: The username of the account.</summary>
        [JsiiProperty(name: "attrUsername", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        public virtual object AttrUsername
        {
            get => GetInstanceProperty<object>()!;
        }

        [JsiiProperty(name: "props", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-rocketmq5.AclProps\"}")]
        public virtual AlibabaCloud.SDK.ROS.CDK.Rocketmq5.IAclProps Props
        {
            get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Rocketmq5.IAclProps>()!;
        }

        [JsiiProperty(name: "enableResourcePropertyConstraint", typeJson: "{\"primitive\":\"boolean\"}")]
        protected virtual bool EnableResourcePropertyConstraint
        {
            get => GetInstanceProperty<bool>()!;
            set => SetInstanceProperty(value);
        }

        [JsiiProperty(name: "id", typeJson: "{\"primitive\":\"string\"}")]
        protected virtual string Id
        {
            get => GetInstanceProperty<string>()!;
            set => SetInstanceProperty(value);
        }

        [JsiiProperty(name: "scope", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.Construct\"}")]
        protected virtual AlibabaCloud.SDK.ROS.CDK.Core.Construct Scope
        {
            get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.Construct>()!;
            set => SetInstanceProperty(value);
        }
    }
}

File: multiple-languages/java/ros-cdk-rocketmq5/src/main/java/com/aliyun/ros/cdk/rocketmq5/Acl.java
package com.aliyun.ros.cdk.rocketmq5;

/**
 * This class encapsulates and extends the ROS resource type <code>ALIYUN::ROCKETMQ5::Acl</code>.
 */
@javax.annotation.Generated(value = "jsii-pacmak/1.85.0 (build 08ee592)", date = "2026-05-07T04:16:53.684Z")
@software.amazon.jsii.Jsii(module = com.aliyun.ros.cdk.rocketmq5.$Module.class, fqn = "@alicloud/ros-cdk-rocketmq5.Acl")
public class Acl extends com.aliyun.ros.cdk.core.Resource implements com.aliyun.ros.cdk.rocketmq5.IAcl {

    protected Acl(final software.amazon.jsii.JsiiObjectRef objRef) {
        super(objRef);
    }

    protected Acl(final software.amazon.jsii.JsiiObject.InitializationMode initializationMode) {
        super(initializationMode);
    }

    /**
     * Param scope - scope in which this resource is defined Param id    - scoped id of the resource Param props - resource properties.
     * <p>
     * @param scope This parameter is required.
     * @param id This parameter is required.
     * @param props This parameter is required.
     * @param enableResourcePropertyConstraint
     */
    public Acl(final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.core.Construct scope, final @org.jetbrains.annotations.NotNull java.lang.String id, final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.rocketmq5.AclProps props, final @org.jetbrains.annotations.Nullable java.lang.Boolean enableResourcePropertyConstraint) {
        super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
        software.amazon.jsii.JsiiEngine.getInstance().createNewObject(this, new Object[] { java.util.Objects.requireNonNull(scope, "scope is required"), java.util.Objects.requireNonNull(id, "id is required"), java.util.Objects.requireNonNull(props, "props is required"), enableResourcePropertyConstraint });
    }

    /**
     * Param scope - scope in which this resource is defined Param id    - scoped id of the resource Param props - resource properties.
     * <p>
     * @param scope This parameter is required.
     * @param id This parameter is required.
     * @param props This parameter is required.
     */
    public Acl(final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.core.Construct scope, final @org.jetbrains.annotations.NotNull java.lang.String id, final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.rocketmq5.AclProps props) {
        super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
        software.amazon.jsii.JsiiEngine.getInstance().createNewObject(this, new Object[] { java.util.Objects.requireNonNull(scope, "scope is required"), java.util.Objects.requireNonNull(id, "id is required"), java.util.Objects.requireNonNull(props, "props is required") });
    }

    /**
     * Attribute Actions: The type of operation authorized.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrActions() {
        return software.amazon.jsii.Kernel.get(this, "attrActions", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    /**
     * Attribute Decision: The decision result of the authorization.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrDecision() {
        return software.amazon.jsii.Kernel.get(this, "attrDecision", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    /**
     * Attribute InstanceId: The ID of the RocketMQ instance.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrInstanceId() {
        return software.amazon.jsii.Kernel.get(this, "attrInstanceId", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    /**
     * Attribute IpWhitelists: The IP addresses in the whitelist.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrIpWhitelists() {
        return software.amazon.jsii.Kernel.get(this, "attrIpWhitelists", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    /**
     * Attribute ResourceName: The name of the resource on which you want to grant permissions.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrResourceName() {
        return software.amazon.jsii.Kernel.get(this, "attrResourceName", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    /**
     * Attribute ResourceType: The type of the resource on which you want to grant permissions.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrResourceType() {
        return software.amazon.jsii.Kernel.get(this, "attrResourceType", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    /**
     * Attribute Username: The username of the account.
     */
    @Override
    public @org.jetbrains.annotations.NotNull java.lang.Object getAttrUsername() {
        return software.amazon.jsii.Kernel.get(this, "attrUsername", software.amazon.jsii.NativeType.forClass(java.lang.Object.class));
    }

    @Override
    public @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.rocketmq5.AclProps getProps() {
        return software.amazon.jsii.Kernel.get(this, "props", software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.rocketmq5.AclProps.class));
    }

    protected @org.jetbrains.annotations.NotNull java.lang.Boolean getEnableResourcePropertyConstraint() {
        return software.amazon.jsii.Kernel.get(this, "enableResourcePropertyConstraint", software.amazon.jsii.NativeType.forClass(java.lang.Boolean.class));
    }

    protected void setEnableResourcePropertyConstraint(final @org.jetbrains.annotations.NotNull java.lang.Boolean value) {
        software.amazon.jsii.Kernel.set(this, "enableResourcePropertyConstraint", java.util.Objects.requireNonNull(value, "enableResourcePropertyConstraint is required"));
    }

    protected @org.jetbrains.annotations.NotNull java.lang.String getId() {
        return software.amazon.jsii.Kernel.get(this, "id", software.amazon.jsii.NativeType.forClass(java.lang.String.class));
    }

    protected void setId(final @org.jetbrains.annotations.NotNull java.lang.String value) {
        software.amazon.jsii.Kernel.set(this, "id", java.util.Objects.requireNonNull(value, "id is required"));
    }

    protected @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.core.Construct getScope() {
        return software.amazon.jsii.Kernel.get(this, "scope", software.amazon.jsii.NativeType.forClass(com.aliyun.ros.cdk.core.Construct.class));
    }

    protected void setScope(final @org.jetbrains.annotations.NotNull com.aliyun.ros.cdk.core.Construct value) {
        software.amazon.jsii.Kernel.set(this, "scope", java.util.Objects.requireNonNull(value, "scope is required"));
    }

    /**
     * A fluent builder for {@link com.aliyun.ros.cdk.rocketmq5.Acl}.
     */
    public static final class Builder implements software.amazon.jsii.Builder<com.aliyun.ros.cdk.rocketmq5.Acl> {
        /**
         * @return a new instance of {@link Builder}.
         * @param scope This parameter is required.
         * @param id This parameter is required.
         * @param enableResourcePropertyConstraint
         */
        public static Builder create(final com.aliyun.ros.cdk.core.Construct scope, final java.lang.String id, final java.lang.Boolean enableResourcePropertyConstraint) {
            return new Builder(scope, id, enableResourcePropertyConstraint);
        }
        /**
         * @return a new instance of {@link Builder}.
         * @param scope This parameter is required.
         * @param id This parameter is required.
         */
        public static Builder create(final com.aliyun.ros.cdk.core.Construct scope, final java.lang.String id) {
            return new Builder(scope, id, null);
        }

        private final com.aliyun.ros.cdk.core.Construct scope;
        private final java.lang.String id;
        private final java.lang.Boolean enableResourcePropertyConstraint;
        private final com.aliyun.ros.cdk.rocketmq5.AclProps.Builder props;

        private Builder(final com.aliyun.ros.cdk.core.Construct scope, final java.lang.String id, final java.lang.Boolean enableResourcePropertyConstraint) {
            this.scope = scope;
            this.id = id;
            this.enableResourcePropertyConstraint = enableResourcePropertyConstraint;
            this.props = new com.aliyun.ros.cdk.rocketmq5.AclProps.Builder();
        }

        /**
         * Property actions: The type of operation authorized.
         * <p>
         * @return {@code this}
         * @param actions Property actions: The type of operation authorized. This parameter is required.
         */
        public Builder actions(final com.aliyun.ros.cdk.core.IResolvable actions) {
            this.props.actions(actions);
            return this;
        }
        /**
         * Property actions: The type of operation authorized.
         * <p>
         * @return {@code this}
         * @param actions Property actions: The type of operation authorized. This parameter is required.
         */
        public Builder actions(final java.util.List<? extends java.lang.Object> actions) {
            this.props.actions(actions);
            return this;
        }

        /**
         * Property decision: The decision result of the authorization.
         * <p>
         * Valid values:
         * <p>
         * <ul>
         * <li>Deny: Access is denied.</li>
         * <li>Allow: Allow access.</li>
         * </ul>
         * <p>
         * @return {@code this}
         * @param decision Property decision: The decision result of the authorization. This parameter is required.
         */
        public Builder decision(final java.lang.String decision) {
            this.props.decision(decision);
            return this;
        }
        /**
         * Property decision: The decision result of the authorization.
         * <p>
         * Valid values:
         * <p>
         * <ul>
         * <li>Deny: Access is denied.</li>
         * <li>Allow: Allow access.</li>
         * </ul>
         * <p>
         * @return {@code this}
         * @param decision Property decision: The decision result of the authorization. This parameter is required.
         */
        public Builder decision(final com.aliyun.ros.cdk.core.IResolvable decision) {
            this.props.decision(decision);
            return this;
        }

        /**
         * Property instanceId: The ID of the RocketMQ instance.
         * <p>
         * @return {@code this}
         * @param instanceId Property instanceId: The ID of the RocketMQ instance. This parameter is required.
         */
        public Builder instanceId(final java.lang.String instanceId) {
            this.props.instanceId(instanceId);
            return this;
        }
        /**
         * Property instanceId: The ID of the RocketMQ instance.
         * <p>
         * @return {@code this}
         * @param instanceId Property instanceId: The ID of the RocketMQ instance. This parameter is required.
         */
        public Builder instanceId(final com.aliyun.ros.cdk.core.IResolvable instanceId) {
            this.props.instanceId(instanceId);
            return this;
        }

        /**
         * Property resourceName: The name of the resource on which you want to grant permissions.
         * <p>
         * @return {@code this}
         * @param resourceName Property resourceName: The name of the resource on which you want to grant permissions. This parameter is required.
         */
        public Builder resourceName(final java.lang.String resourceName) {
            this.props.resourceName(resourceName);
            return this;
        }
        /**
         * Property resourceName: The name of the resource on which you want to grant permissions.
         * <p>
         * @return {@code this}
         * @param resourceName Property resourceName: The name of the resource on which you want to grant permissions. This parameter is required.
         */
        public Builder resourceName(final com.aliyun.ros.cdk.core.IResolvable resourceName) {
            this.props.resourceName(resourceName);
            return this;
        }

        /**
         * Property resourceType: The type of the resource on which you want to grant permissions.
         * <p>
         * Valid values:
         * <p>
         * <ul>
         * <li>Group: consumer Group.</li>
         * <li>Topic: Topic.</li>
         * </ul>
         * <p>
         * @return {@code this}
         * @param resourceType Property resourceType: The type of the resource on which you want to grant permissions. This parameter is required.
         */
        public Builder resourceType(final java.lang.String resourceType) {
            this.props.resourceType(resourceType);
            return this;
        }
        /**
         * Property resourceType: The type of the resource on which you want to grant permissions.
         * <p>
         * Valid values:
         * <p>
         * <ul>
         * <li>Group: consumer Group.</li>
         * <li>Topic: Topic.</li>
         * </ul>
         * <p>
         * @return {@code this}
         * @param resourceType Property resourceType: The type of the resource on which you want to grant permissions. This parameter is required.
         */
        public Builder resourceType(final com.aliyun.ros.cdk.core.IResolvable resourceType) {
            this.props.resourceType(resourceType);
            return this;
        }

        /**
         * Property username: The username of the account.
         * <p>
         * @return {@code this}
         * @param username Property username: The username of the account. This parameter is required.
         */
        public Builder username(final java.lang.String username) {
            this.props.username(username);
            return this;
        }
        /**
         * Property username: The username of the account.
         * <p>
         * @return {@code this}
         * @param username Property username: The username of the account. This parameter is required.
         */
        public Builder username(final com.aliyun.ros.cdk.core.IResolvable username) {
            this.props.username(username);
            return this;
        }

        /**
         * Property ipWhitelists: The IP addresses in the whitelist.
         * <p>
         * @return {@code this}
         * @param ipWhitelists Property ipWhitelists: The IP addresses in the whitelist. This parameter is required.
         */
        public Builder ipWhitelists(final com.aliyun.ros.cdk.core.IResolvable ipWhitelists) {
            this.props.ipWhitelists(ipWhitelists);
            return this;
        }
        /**
         * Property ipWhitelists: The IP addresses in the whitelist.
         * <p>
         * @return {@code this}
         * @param ipWhitelists Property ipWhitelists: The IP addresses in the whitelist. This parameter is required.
         */
        public Builder ipWhitelists(final java.util.List<? extends java.lang.Object> ipWhitelists) {
            this.props.ipWhitelists(ipWhitelists);
            return this;
        }

        /**
         * @return a newly built instance of {@link com.aliyun.ros.cdk.rocketmq5.Acl}.
         */
        @Override
        public com.aliyun.ros.cdk.rocketmq5.Acl build() {
            return new com.aliyun.ros.cdk.rocketmq5.Acl(
                this.scope,
                this.id,
                this.props.build(),
                this.enableResourcePropertyConstraint
            );
        }
    }
}
